### PR TITLE
[WIP] resolve image secrets in the buildcontroller instead of when generating the build

### PR DESCRIPTION
**Pull request by bparees**
need to relocate some tests that ran in the generator before and convince myself we have adequate extended test coverage for these secrets.

fixes https://github.com/openshift/origin/issues/16609

### CODE REVIEW

File: pkg/build/controller/build/build_controller.go [hunk 1]
```diff
@@ -789,6 +789,61 @@ func (bc *BuildController) createBuildPod(build *buildapi.Build) (*buildUpdate,
 	}
 	build.Spec.Output.PushSecret = pushSecret
 
+	// Set the pushSecret that will be needed by the build to push the image to the registry
```
Comment (csrwng) [inline pkg/build/controller/build/build_controller.go]: pull secret

File: pkg/build/controller/build/build_controller.go [hunk 2]
```diff
@@ -789,6 +789,61 @@ func (bc *BuildController) createBuildPod(build *buildapi.Build) (*buildUpdate,
...
+	case build.Spec.Strategy.DockerStrategy != nil:
+		pullSecret = build.Spec.Strategy.DockerStrategy.PullSecret
+		if build.Spec.Strategy.DockerStrategy.From != nil {
+			imageName = build.Spec.Strategy.DockerStrategy.From.Name
+		}
+	case build.Spec.Strategy.CustomStrategy != nil:
+		pullSecret = build.Spec.Strategy.CustomStrategy.PullSecret
+		imageName = build.Spec.Strategy.CustomStrategy.From.Name
+	}
+	// Only look up a push secret if the user hasn't explicitly provided one.
```
Comment (csrwng) [inline pkg/build/controller/build/build_controller.go]: pull secret

File: pkg/build/controller/build/build_controller.go (hunk 1 above)
Comment (bparees) [reply pkg/build/controller/build/build_controller.go]: thanks

**Comment by bparees** [comment]
> In general looks good to me, just minor comments below. Also, we should add a new integration test to test the build controller behavior (as far as resolving images/ setting secrets).

yeah, that's what this was about:

> need to relocate some tests that ran in the generator before and convince myself we have adequate extended test coverage for these secrets.


That said, the extended test that's currently failing already demonstrates we have some coverage for this today, at least for image source inputs that require a secret.  I have mixed feelings about duplicating that in an intergration test.  Will explore that further after i get things fully working.

File: pkg/api/meta/builds.go [hunk 3]
```diff
@@ -24,8 +24,8 @@ func NewBuildMutator(build *buildapi.Build) ImageReferenceMutator {
 
 func (m *buildSpecMutator) Mutate(fn ImageReferenceMutateFunc) field.ErrorList {
 	var errs field.ErrorList
-	for i, image := range m.spec.Source.Images {
-		if err := fn(&image.From); err != nil {
+	for i := range m.spec.Source.Images {
+		if err := fn(&m.spec.Source.Images[i].From); err != nil {
```
Comment (bparees) [inline pkg/api/meta/builds.go]: @smarterclayton @csrwng fyi, this was never doing what it was supposed to do before, it was only mutating the copy of image that the range iterator was making.  yay go.

**Comment by bparees** [comment]
moving this work to here: https://github.com/openshift/origin/pull/16737

**Pull request closed without merging**